### PR TITLE
migrate services to nullsafety

### PR DESCRIPTION
**Pull request by a14n**
## Description

NNBD migration for services

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/tools/gen_keycodes/data/keyboard_key.tmpl
Comment (goderbauer) [inline dev/tools/gen_keycodes/data/keyboard_key.tmpl:164]: /cc @dkwingsmt or @gspencergoog Is it expected that the label can be null? If so, can we add something to the docs here to explain what that means?

File: packages/flutter/lib/src/services/autofill.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/autofill.dart:695]: /cc @LongCatIsLooong When is this expected to be null? Can we add that to the documentation here?

File: packages/flutter/lib/src/services/platform_views.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/platform_views.dart:59]: Should this be `late` to keep the laziness of the original code?

**Review by goderbauer — COMMENTED** [review]
I've reviewed about half of this, taking a break since it is so giant.

File: packages/flutter/lib/src/services/autofill.dart
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/autofill.dart:695]: It's not supposed to be null but I might have omitted `currentEditingValue` in some of the tests.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (a14n) [reply packages/flutter/lib/src/services/platform_views.dart:59]: Static variables and top level variables are already lazy in Dart so I think `static late final` is the same as `static final`.
@leafpetersen or @lrhn is it correct? 

If yes the analyzer could perhaps have an hint to prefer `static final` over `static late final`

File: packages/flutter/lib/src/services/autofill.dart
Comment (a14n) [reply packages/flutter/lib/src/services/autofill.dart:695]: By making `currentEditingValue` required and non-nullable tests on packages/flutter are OK.

I made it nullable because in the constructor it was not `@required` and not assered as not null.

I'll make the change to have non-nullable type.

**Comment by a14n** [comment]
>I've reviewed about half of this, taking a break since it is so giant.

If I continue PR by library there will be longer PR than this one. Should I make smaller PR (for instance with ~10 files) to simplify your review? I just hope to not face mixed language versions issues.

File: dev/tools/gen_keycodes/data/keyboard_key.tmpl
Comment (a14n) [reply dev/tools/gen_keycodes/data/keyboard_key.tmpl:164]: null seems to be possible regarding https://github.com/flutter/flutter/blob/7ec2d360bce51f010554be8d80572af4748a4a95/dev/tools/gen_keycodes/lib/keyboard_keys_code_gen.dart#L61-L73

Comment (gspencergoog) [reply dev/tools/gen_keycodes/data/keyboard_key.tmpl:164]: Yes, null is possible, since it's possible that we don't have the key label data for a key.

Comment (goderbauer) [reply dev/tools/gen_keycodes/data/keyboard_key.tmpl:164]: Can we update the doc comment here to mention that?

**Comment by goderbauer** [comment]
@a14n Smaller PRs are preferred as they are faster to review, but if it is not possible, I am also OK with just reviewing the larger ones in chunks.

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart:597]: This is odd. Returning null from here is always a bug, it means we missed a case in the switch. This should really be non-nullable (unless null is a legal value for affinity).

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart:665]: The `?? TextAffinity.downstream` is strange since `_toTextAffinity` effectively can't return null (see comment above).

File: packages/flutter/lib/src/services/system_channels.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/system_channels.dart:220]: This surprises me, why is null allowed here? (according to the doc it only sends string representation of enum values.)

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [reply packages/flutter/lib/src/services/text_input.dart:665]: Maybe `encoded['selectionAffinity']` can actually be null?

**Review by goderbauer — COMMENTED** [review]
I have one file left to review, but had to run to a meeting. Will review when I get back.

**Review by goderbauer — COMMENTED** [review]
Reveiwed all.

File: dev/tools/gen_keycodes/data/keyboard_key.tmpl
Comment (a14n) [reply dev/tools/gen_keycodes/data/keyboard_key.tmpl:166]: Done

File: packages/flutter/lib/src/services/binding.dart
Comment (a14n) [reply packages/flutter/lib/src/services/binding.dart]: Changed to non-nullable

File: packages/flutter/lib/src/services/system_channels.dart
Comment (a14n) [reply packages/flutter/lib/src/services/system_channels.dart:220]: Because `StringCodec` implements `MessageCodec<String?>` and its decode/encode methods accepts/returns null.

File: packages/flutter/lib/src/services/text_input.dart
Comment (a14n) [reply packages/flutter/lib/src/services/text_input.dart]: Changed

File: packages/flutter/lib/src/services/text_input.dart
Comment (a14n) [reply packages/flutter/lib/src/services/text_input.dart]: No idea :-( Let's run tests and see...

File: packages/flutter/lib/src/services/binding.dart
Comment (a14n) [reply packages/flutter/lib/src/services/binding.dart]: Thanks to CI (https://cirrus-ci.com/task/5831088616505344?command=main#L38) we have an answer:

```
00:00 +0: loading /private/tmp/flutter sdk/dev/automated_tests/test_smoke_test/timeout_pass_test.dart                                                                                                  
00:01 +0: loading /private/tmp/flutter sdk/dev/automated_tests/test_smoke_test/timeout_pass_test.dart                                                                                                  
00:02 +0: loading /private/tmp/flutter sdk/dev/automated_tests/test_smoke_test/timeout_pass_test.dart                                                                                                  
00:02 +0 -1: loading /private/tmp/flutter sdk/dev/automated_tests/test_smoke_test/timeout_pass_test.dart [E]                                                                                           
  Failed to load "/private/tmp/flutter sdk/dev/automated_tests/test_smoke_test/timeout_pass_test.dart": Invalid argument(s): AppLifecycleState null not supported.
  package:flutter/src/services/binding.dart 207:5     ServicesBinding._parseAppLifecycleMessage
  package:flutter/src/services/binding.dart 185:38    ServicesBinding.readInitialLifecycleStateFromNativeWindow
  package:flutter/src/services/binding.dart 35:5      ServicesBinding.initInstances
  package:flutter/src/gestures/binding.dart 65:11     GestureBinding.initInstances
  package:flutter/src/semantics/binding.dart 24:11    SemanticsBinding.initInstances
  package:flutter/src/rendering/binding.dart 32:11    RendererBinding.initInstances
  package:flutter/src/painting/binding.dart 23:11     PaintingBinding.initInstances
  package:flutter/src/widgets/binding.dart 257:11     WidgetsBinding.initInstances
  package:flutter_test/src/binding.dart 318:11        TestWidgetsFlutterBinding.initInstances
  package:flutter_test/src/binding.dart 914:11        AutomatedTestWidgetsFlutterBinding.initInstances
```

File: packages/flutter/lib/src/services/text_input.dart
Comment (goderbauer) [inline packages/flutter/lib/src/services/text_input.dart:665]: Looking at this code again, I think the idea is that if `encoded['selectionAffinity']` is null, this should default to `TextAffinity.downstream`.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/services/platform_views.dart
Comment (leafpetersen) [reply packages/flutter/lib/src/services/platform_views.dart:59]: `static late final` with an initializer is the same as `static final` (and also when non-final).  Without an initializer they are different.  +1 for a hint/lint.  cc @bwilkerson

**Comment by a14n** [comment]
@goderbauer I see _Google testing failed. Please contact a Google developer to investigate._ could you check what is wrong?

**Comment by gspencergoog** [comment]
Michael is out today, I'll check on it.

**Comment by gspencergoog** [comment]
The Google internal failure looks like a failure of the infrastructure that tests the Flutter PRs, not anything specific with this PR.

**Comment by a14n** [comment]
@gspencergoog do you know why flutter-gold is still _pending_ after 10 hours?

**Comment by a14n** [comment]
>The Google internal failure looks like a failure of the infrastructure that tests the Flutter PRs, not anything specific with this PR.

Does it mean I can merge this PR even if _Google testing_ job fails?

**Comment by gspencergoog** [comment]
Yes, you can merge in this case even if the Google tests are failing, since we've determined that it hasn't got to do with this PR.

As for the Gold delay, I'm not sure: it doesn't appear to need any triage, so I'm not sure how to clear it.  I'll ask.

**Comment by Piinks** [comment]
> As for the Gold delay, I'm not sure: it doesn't appear to need any triage, so I'm not sure how to clear it. I'll ask.

The gold check is currently broken, see #62682 
The merge bot will be blocked by it, so this will have to be landed manually.

**Comment by gspencergoog** [comment]
OK, looks like we're good to land this (manually, not with the "waiting for tree to go green" bot). Do it quickly, while the tree is green... :-)

**Comment by creativecreatorormaybenot** [comment]
I think this causes text input on web with null safety to fail (#62724).

File: packages/flutter/lib/src/services/text_input.dart
Comment (a14n) [reply packages/flutter/lib/src/services/text_input.dart:665]: Actually this is null on web plateform and is the cause of #62724.

**Pull request merged**
